File: packages/snow_draw_core/lib/draw/types/edit_context.dart
import 'package:meta/meta.dart';

import '../models/element_state.dart';
import '../services/text/text_metrics_service.dart';
import 'draw_point.dart';
import 'draw_rect.dart';
import 'element_geometry.dart';
import 'resize_mode.dart';

/// Immutable context snapshot for an edit operation.
///
/// Created on edit-start and kept unchanged for the entire edit session.
///
/// Memory note:
/// Each subclass now carries minimal geometry snapshots (ElementMoveSnapshot,
/// ElementResizeSnapshot, ElementRotateSnapshot) instead of full ElementState.
@immutable
abstract class EditContext {
  const EditContext({
    required this.startPosition,
    required this.startBounds,
    required this.selectedIdsAtStart,
    required this.selectionVersion,
    required this.elementsVersion,
  });

  /// Pointer position at the start of the edit operation (world coordinates).
  final DrawPoint startPosition;

  /// Selection overlay bounds at the start of the edit operation.
  final DrawRect startBounds;

  /// Selected element ids at the start of the edit operation.
  ///
  /// This snapshot is part of the edit session invariants:
  /// - preview should be based on this set
  /// - finish/commit must not depend on selection changing during editing
  final Set<String> selectedIdsAtStart;

  /// Selection version captured when the edit session started.
  final int selectionVersion;

  /// Elements version captured when the edit session started.
  final int elementsVersion;

  DrawPoint get startCenter => startBounds.center;

  bool get isSingleSelect => selectedIdsAtStart.length == 1;
  bool get isMultiSelect => selectedIdsAtStart.length > 1;

  /// Whether this context carries non-empty element snapshots.
  bool get hasSnapshots => false;
}

/// Context for move operations.
///
/// Uses ElementMoveSnapshot to store each element's starting center,
/// saving about 92% memory compared to full ElementState.
@immutable
final class MoveEditContext extends EditContext {
  const MoveEditContext({
    required super.startPosition,
    required super.startBounds,
    required super.selectedIdsAtStart,
    required super.selectionVersion,
    required super.elementsVersion,
    required this.elementSnapshots,
    this.snapBoundsAtStart,
    this.referenceElements = const [],
    this.referenceElementAabbs = const [],
  });

  /// Starting center for each element (lean snapshot).
  ///
  /// Stores only the data needed for move operations: element centers.
  final Map<String, ElementMoveSnapshot> elementSnapshots;

  /// Selection bounds resolved from selected element geometry at edit start.
  ///
  /// This is used as the snapping base so move snapping stays deterministic
  /// even when document geometry changes during the drag.
  final DrawRect? snapBoundsAtStart;

  /// Non-selected visible elements captured at edit start for object snapping.
  final List<ElementState> referenceElements;

  /// Precomputed axis-aligned bounds for [referenceElements].
  final List<DrawRect> referenceElementAabbs;

  /// Bounds used as the snap base for this move session.
  DrawRect get snapBounds => snapBoundsAtStart ?? startBounds;

  @override
  bool get hasSnapshots => elementSnapshots.isNotEmpty;
}

/// Context for resize operations.
///
/// Uses ElementResizeSnapshot to store each element's starting rect and
/// rotation, saving about 80% memory compared to full ElementState.
@immutable
final class ResizeEditContext extends EditContext {
  const ResizeEditContext({
    required super.startPosition,
    required super.startBounds,
    required super.selectedIdsAtStart,
    required super.selectionVersion,
    required super.elementsVersion,
    required this.resizeMode,
    required this.handleOffset,
    required this.rotation,
    required this.elementSnapshots,
    this.selectionPadding = 0.0,
    this.referenceElements = const [],
    this.referenceElementAabbs = const [],
    this.forceSerialNumberAspectRatio = false,
    this.textMetricsService = defaultTextMetricsService,
  });
  final ResizeMode resizeMode;
  final DrawPoint handleOffset;
  final double rotation;
  final double selectionPadding;

  /// Starting geometry for each element (lean snapshot).
  final Map<String, ElementResizeSnapshot> elementSnapshots;

  /// Non-selected visible elements captured at edit start for object snapping.
  final List<ElementState> referenceElements;

  /// Precomputed axis-aligned bounds for [referenceElements].
  final List<DrawRect> referenceElementAabbs;

  /// Whether resize should always preserve aspect ratio for this session.
  final bool forceSerialNumberAspectRatio;

  /// Text metrics service used by text resize layout calculations.
  final TextMetricsService textMetricsService;

  @override
  bool get hasSnapshots => elementSnapshots.isNotEmpty;

  bool get hasRotation => rotation != 0;

  ResizeEditContext withTextMetricsService(
    TextMetricsService textMetricsService,
  ) {
    if (identical(this.textMetricsService, textMetricsService)) {
      return this;
    }
    return ResizeEditContext(
      startPosition: startPosition,
      startBounds: startBounds,
      selectedIdsAtStart: selectedIdsAtStart,
      selectionVersion: selectionVersion,
      elementsVersion: elementsVersion,
      resizeMode: resizeMode,
      handleOffset: handleOffset,
      rotation: rotation,
      selectionPadding: selectionPadding,
      elementSnapshots: elementSnapshots,
      referenceElements: referenceElements,
      referenceElementAabbs: referenceElementAabbs,
      forceSerialNumberAspectRatio: forceSerialNumberAspectRatio,
      textMetricsService: textMetricsService,
    );
  }
}

/// Context for rotate operations.
///
/// Uses ElementRotateSnapshot to store each element's starting center and
/// rotation, saving about 88% memory compared to full ElementState.
@immutable
final class RotateEditContext extends EditContext {
  const RotateEditContext({
    required super.startPosition,
    required super.startBounds,
    required super.selectedIdsAtStart,
    required super.selectionVersion,
    required super.elementsVersion,
    required this.startAngle,
    required this.baseRotation,
    required this.rotationSnapAngle,
    required this.elementSnapshots,
  });

  /// Pointer angle around [startCenter] at start (raw atan2 angle).
  final double startAngle;

  /// Base rotation at start (single-select: element rotation; multi-select:
  /// overlay rotation).
  final double baseRotation;

  /// Discrete snap interval for rotation (radians).
  final double rotationSnapAngle;

  /// Starting rotation info for each element (lean snapshot).
  final Map<String, ElementRotateSnapshot> elementSnapshots;

  @override
  bool get hasSnapshots => elementSnapshots.isNotEmpty;
}

File: packages/snow_draw_rust_engine/src/draw/types/edit_context.rs
#![allow(dead_code)]

use crate::draw::elements::types::text::text_editing_geometry::{
    FallbackTextMetricsService as GeometryFallbackTextMetricsService,
    TextLayoutRequest as GeometryTextLayoutRequest, TextMetrics as GeometryTextMetrics,
    TextMetricsService as GeometryTextMetricsService,
};
use crate::draw::services::text::text_metrics_service::{
    FallbackTextMetricsService as ResizeFallbackTextMetricsService,
    TextMetricsService as ResizeTextMetricsService,
};
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;
use crate::draw::types::element_geometry::{
    ElementMoveSnapshot, ElementResizeSnapshot, ElementRotateSnapshot,
};
use crate::draw::types::resize_mode::ResizeMode;
use std::collections::{HashMap, HashSet};
use std::fmt;
use std::sync::Arc;

/// Shared base context captured at edit-start.
///
/// This snapshot is immutable for the lifetime of an edit session and
/// represents invariants that preview and commit logic should rely on.
#[derive(Clone, Debug)]
pub struct EditContext {
    /// Pointer position at the start of the edit operation (world coordinates).
    pub start_position: DrawPoint,

    /// Selection overlay bounds at the start of the edit operation.
    pub start_bounds: DrawRect,

    /// Selected element ids at the start of the edit operation.
    pub selected_ids_at_start: HashSet<String>,

    /// Selected element ids in the observable iteration order captured at edit start.
    pub selected_ids_at_start_in_order: Vec<String>,

    /// Selection version captured when the edit session started.
    pub selection_version: i64,

    /// Elements version captured when the edit session started.
    pub elements_version: i64,
}

impl EditContext {
    pub fn new(
        start_position: DrawPoint,
        start_bounds: DrawRect,
        selected_ids_at_start: HashSet<String>,
        selection_version: i64,
        elements_version: i64,
    ) -> Self {
        let selected_ids_at_start_in_order = selected_ids_at_start.iter().cloned().collect();
        Self::new_with_order(
            start_position,
            start_bounds,
            selected_ids_at_start,
            selected_ids_at_start_in_order,
            selection_version,
            elements_version,
        )
    }

    pub fn new_with_order(
        start_position: DrawPoint,
        start_bounds: DrawRect,
        selected_ids_at_start: HashSet<String>,
        selected_ids_at_start_in_order: Vec<String>,
        selection_version: i64,
        elements_version: i64,
    ) -> Self {
        Self {
            start_position,
            start_bounds,
            selected_ids_at_start,
            selected_ids_at_start_in_order,
            selection_version,
            elements_version,
        }
    }

    pub fn start_center(&self) -> DrawPoint {
        self.start_bounds.center()
    }

    pub fn is_single_select(&self) -> bool {
        self.selected_ids_at_start.len() == 1
    }

    pub fn is_multi_select(&self) -> bool {
        self.selected_ids_at_start.len() > 1
    }

    pub fn selected_ids_at_start_in_order(&self) -> &[String] {
        &self.selected_ids_at_start_in_order
    }
}

/// Behavior shared by concrete edit-context kinds.
pub trait EditContextLike {
    fn base(&self) -> &EditContext;

    fn start_center(&self) -> DrawPoint {
        self.base().start_center()
    }

    fn is_single_select(&self) -> bool {
        self.base().is_single_select()
    }

    fn is_multi_select(&self) -> bool {
        self.base().is_multi_select()
    }

    fn has_snapshots(&self) -> bool {
        false
    }
}

impl EditContextLike for EditContext {
    fn base(&self) -> &EditContext {
        self
    }
}

/// Reference element snapshot used by snapping contexts.
pub type ElementState = crate::draw::models::element_state::ElementState;

/// Context for move operations.
///
/// Stores only per-element centers using [`ElementMoveSnapshot`] to keep the
/// session snapshot lean.
#[derive(Clone, Debug)]
pub struct MoveEditContext {
    pub base: EditContext,
    pub element_snapshots: HashMap<String, ElementMoveSnapshot>,
    pub snap_bounds_at_start: Option<DrawRect>,
    pub reference_elements: Vec<ElementState>,
    pub reference_element_aabbs: Vec<DrawRect>,
}

impl MoveEditContext {
    pub fn new(base: EditContext, element_snapshots: HashMap<String, ElementMoveSnapshot>) -> Self {
        Self {
            base,
            element_snapshots,
            snap_bounds_at_start: None,
            reference_elements: Vec::new(),
            reference_element_aabbs: Vec::new(),
        }
    }

    pub fn snap_bounds(&self) -> DrawRect {
        self.snap_bounds_at_start.unwrap_or(self.base.start_bounds)
    }
}

impl EditContextLike for MoveEditContext {
    fn base(&self) -> &EditContext {
        &self.base
    }

    fn has_snapshots(&self) -> bool {
        !self.element_snapshots.is_empty()
    }
}

/// Text metrics abstraction used by resize calculations.
///
/// The same service is also consulted by text-editing reducers. Implementers
/// can optionally expose a text-editing compatible view.
pub trait TextMetricsService: fmt::Debug + Send + Sync {
    fn as_text_editing_metrics_service(&self) -> Option<&dyn GeometryTextMetricsService> {
        None
    }

    fn as_text_resize_metrics_service(&self) -> Option<&dyn ResizeTextMetricsService> {
        None
    }
}

#[derive(Debug, Default)]
pub struct DefaultTextMetricsService;

impl TextMetricsService for DefaultTextMetricsService {
    fn as_text_editing_metrics_service(&self) -> Option<&dyn GeometryTextMetricsService> {
        Some(self)
    }

    fn as_text_resize_metrics_service(&self) -> Option<&dyn ResizeTextMetricsService> {
        Some(self)
    }
}

impl GeometryTextMetricsService for DefaultTextMetricsService {
    fn measure(&self, request: &GeometryTextLayoutRequest<'_>) -> GeometryTextMetrics {
        let fallback = GeometryFallbackTextMetricsService;
        fallback.measure(request)
    }
}

impl ResizeTextMetricsService for DefaultTextMetricsService {
    fn measure(
        &self,
        request: &crate::draw::services::text::text_metrics_service::TextLayoutRequest<'_>,
    ) -> crate::draw::services::text::text_metrics_service::TextMetrics {
        let fallback = ResizeFallbackTextMetricsService;
        fallback.measure(request)
    }
}

/// Returns the default text metrics implementation for resize sessions.
pub fn default_text_metrics_service() -> Arc<dyn TextMetricsService> {
    Arc::new(DefaultTextMetricsService)
}

/// Context for resize operations.
///
/// Stores the minimum geometry needed for deterministic resize behavior across
/// the full drag session.
#[derive(Clone, Debug)]
pub struct ResizeEditContext {
    pub base: EditContext,
    pub resize_mode: ResizeMode,
    pub handle_offset: DrawPoint,
    pub rotation: f64,
    pub selection_padding: f64,
    pub element_snapshots: HashMap<String, ElementResizeSnapshot>,
    pub reference_elements: Vec<ElementState>,
    pub reference_element_aabbs: Vec<DrawRect>,
    pub force_serial_number_aspect_ratio: bool,
    pub text_metrics_service: Arc<dyn TextMetricsService>,
}

impl ResizeEditContext {
    pub fn new(
        base: EditContext,
        resize_mode: ResizeMode,
        handle_offset: DrawPoint,
        rotation: f64,
        element_snapshots: HashMap<String, ElementResizeSnapshot>,
    ) -> Self {
        Self {
            base,
            resize_mode,
            handle_offset,
            rotation,
            selection_padding: 0.0,
            element_snapshots,
            reference_elements: Vec::new(),
            reference_element_aabbs: Vec::new(),
            force_serial_number_aspect_ratio: false,
            text_metrics_service: default_text_metrics_service(),
        }
    }

    pub fn has_rotation(&self) -> bool {
        self.rotation != 0.0
    }

    pub fn with_text_metrics_service(
        &self,
        text_metrics_service: Arc<dyn TextMetricsService>,
    ) -> Self {
        if Arc::ptr_eq(&self.text_metrics_service, &text_metrics_service) {
            return self.clone();
        }

        let mut updated = self.clone();
        updated.text_metrics_service = text_metrics_service;
        updated
    }
}

impl EditContextLike for ResizeEditContext {
    fn base(&self) -> &EditContext {
        &self.base
    }

    fn has_snapshots(&self) -> bool {
        !self.element_snapshots.is_empty()
    }
}

/// Context for rotate operations.
///
/// Captures initial pointer angle, base rotation and per-element rotation
/// snapshots used to compute deterministic rotation deltas.
#[derive(Clone, Debug)]
pub struct RotateEditContext {
    pub base: EditContext,
    pub start_angle: f64,
    pub base_rotation: f64,
    pub rotation_snap_angle: f64,
    pub element_snapshots: HashMap<String, ElementRotateSnapshot>,
}

impl RotateEditContext {
    pub fn new(
        base: EditContext,
        start_angle: f64,
        base_rotation: f64,
        rotation_snap_angle: f64,
        element_snapshots: HashMap<String, ElementRotateSnapshot>,
    ) -> Self {
        Self {
            base,
            start_angle,
            base_rotation,
            rotation_snap_angle,
            element_snapshots,
        }
    }
}

impl EditContextLike for RotateEditContext {
    fn base(&self) -> &EditContext {
        &self.base
    }

    fn has_snapshots(&self) -> bool {
        !self.element_snapshots.is_empty()
    }
}
